Tests/Realm.Tests/Database/AsyncTests.cs

```diff
@@ -155,5 +158,405 @@ static async Task<long> MeasureTiming(Func<Task> func)
+            });
+        }
+
+        [Test]
+        public void AsyncBeginWrite_RollBack_DoesNot_PersistData()
+        {
+            TestHelpers.RunAsyncTest(async () =>
+            {
+                var personName = "Jonh Wickie";
+
+                using (var transaction = await _realm.BeginWriteAsync())
+                {
+                    var person = new Person { FullName = personName };
+                    _realm.Add(person);
+                    transaction.Rollback();
+                }
+
+                Assert.That(_realm.All<Person>().Count(), Is.EqualTo(0));
+            });
+        }
+
+        [Test]
+        public void AsyncBeginWriteAndClose_DoesNot_PersistData()
+        {
+            TestHelpers.RunAsyncTest(async () =>
+            {
+                IQueryable<Person> peopleQuery = null;
+
+                using (var innerRealm = GetRealm(_realm.Config))
+                {
+                    peopleQuery = innerRealm.All<Person>();
+                    Assert.That(innerRealm.All<Person>().Count(), Is.EqualTo(0));
+
+                    var personName = "Jonh Wickie";
+                    Assert.That(peopleQuery.Count(), Is.EqualTo(0));
+
+                    using (var transaction = await innerRealm.BeginWriteAsync())
+                    {
+                        var person = new Person { FullName = personName };
+                        innerRealm.Add(person);
+                        Assert.That(peopleQuery.Count(), Is.EqualTo(1));
+                    }
+                }
+
+                Assert.That(_realm.All<Person>().Count(), Is.EqualTo(0));
+            });
+        }
+
+        [Test]
+        public void WriteAsync_Start_Multiple_AsyncTransaction_NoDeadlock()
+        {
+            TestHelpers.RunAsyncTest(async () =>
+            {
+                var transaction = await _realm.BeginWriteAsync();
+
+                try
+                {
+                    _realm.Add(new Person
+                    {
+                        FirstName = "Marco"
+                    });
+
+                    var writeTask = _realm.WriteAsync(() =>
+                    {
+                        _realm.Add(new Person
+                        {
+                            FirstName = "Giovanni"
+                        });
+                    });
+
+                    await transaction.CommitAsync();
+                    await writeTask;
+
+                    var peopleQuery = _realm.All<Person>();
+                    Assert.That(peopleQuery.Count, Is.EqualTo(2));
+                    Assert.That(peopleQuery.ElementAt(0).FirstName, Is.EqualTo("Marco"));
+                    Assert.That(peopleQuery.ElementAt(1).FirstName, Is.EqualTo("Giovanni"));
+                }
+                catch
+                {
+                    transaction.Dispose();
+                }
+            });
+        }
+
+        [Test]
+        public void WriteAsync_CaptureObjectsInDelegate_DoesNotThrow()
```

Comment:
Not sure I understand the purpose of this test - what could happen for this to ever throw?

Reply:
This would have worked previously as well because your person instance is unmanaged and can travel between threads freely.

Reply:
I see the mistake. I'll make that instance managed, so the test becomes meaningful. Thanks.